python_modules/dagster/dagster/_utils/cached_method.py

```diff
@@ -0,0 +1,110 @@
+from functools import wraps
+
+from dagster import _check as check
+
+NO_VALUE_IN_CACHE_SENTINEL = object()
+
+
+def cached_method(method):
+    """
+    Caches the results of a method call.
+
+    Usage:
+
+        .. code-block:: python
+
+            class MyClass:
+                @cached_method
+                def fetch_value_from_database(self, key):
+                    ...
+
+    The main difference between this decorator and `functools.lru_cache(max_size=None)` is that each
+    instance of the class whose method is decorated gets its own cache. This means that the cache
+    can be garbage-collected when the object is garbage-collected.
+
+    A more subtle difference between this decorator and `functools.lru_cache` is that this one
+    prioritizes preventing mistakes over ease-of-use and performance. With `functools.lru_cache`,
+    these three invocations would all result in separate cache entries:
+
+        .. code-block:: python
+
+            class MyClass:
+                @cached_method
+                def a_method(self, arg1, arg2):
+                    ...
+
+            obj = MyClass()
+            obj.a_method(arg1="a", arg2=5)
+            obj.a_method(arg2=5, arg1="a")
+            obj.a_method("a", 5)
+
+    With this decorator, the first two would point to the same cache entry, and non-kwarg arguments
+    are not allowed.
+    """
+
+    @wraps(method)
+    def helper(self, *args, **kwargs):
+        cache_attr_name = method.__name__ + "_cache"
+        if not hasattr(self, cache_attr_name):
+            cache = {}
+            setattr(self, cache_attr_name, cache)
+        else:
+            cache = getattr(self, cache_attr_name)
+
+        key = _make_key(args, kwargs)
+        cached_result = cache.get(key, NO_VALUE_IN_CACHE_SENTINEL)
+        if cached_result is not NO_VALUE_IN_CACHE_SENTINEL:
+            return cached_result
+        else:
+            result = method(self, *args, **kwargs)
+            cache[key] = result
+            return result
+
+    return helper
```

Comment:
could be nice to have the `cache_info` setup from lru_cache as well for debugging hits/misses